File: Data Structures & Algorithms/house-robber/submission-1.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        dp = [0] * len(nums)
        if len(nums) == 0:
            return 0

        if len(nums) == 1:
            return nums[0]

        # dp[0] = nums[0]
        # dp[1] = max(nums[0], nums[1])
        # dp[2] = max(dp[1], nums[0] + nums[2])

        dp[0] = nums[0]
        dp[1] = max(nums[0], nums[1])

        for i in range(2, len(nums)):
            dp[i] = max(dp[i - 1], dp[i - 2] + nums[i])

        return dp[len(nums) - 1]

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-1.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        if (s.length() == 0) {
            return 0;
        } else if (s.length() == 1) {
            return 1;
        }

        Set<Character> uniqueChars = new HashSet<Character>();
        int left = 0;
        int maxSubstring = 0;

        for (int right = 0; right < s.length(); right++) {
            while (uniqueChars.contains(s.charAt(right))) {
                // Removing the left char if it is seen
                uniqueChars.remove(s.charAt(left));
                left++;
            }
            // Need to make sure we add the right char each time
            uniqueChars.add(s.charAt(right));
            maxSubstring = Math.max(maxSubstring, right - left + 1);
        }

        return maxSubstring;
    }
}

// pwwkew
// left = 1
// right = 3
// maxSubstring = 2
